Make kMaxStackAlloca

SebastianBoblestETAS opened the pull request:
I added the option tir.max_stack_alloca to overwrite the default value 1024 of kMaxStackAlloca

junrushao1994 commented:
Thanks for the contribution! I don't have much experience tweaking this constant, but by grepping in the codebase, it doesn't seem to be used anywhere other than `lower_tvm_buildin.cc`. Therefore the PR does make sense to me. Shall we also remove `kMaxStackAlloca` in `include/tvm/runtime/device_api.h` to avoid misuse? Thanks!

SebastianBoblestETAS commented:
@junrushao1994 Thanks for the comment. Makes total sense. I removed `kMaxStackAlloca` from `device_api.h` and added it as `kMaxStackAllocaDefault` in `lower_tvm_buildin.cc.`

SebastianBoblestETAS commented:
So, one of the checks fails. I looked at the error message but I do not see any connection to the changes I made.
How should I proceed?

kparzysz-quic commented on src/tir/transforms/lower_tvm_builtin.cc:
```diff
@@ -127,7 +134,7 @@ class BuiltinLower : public StmtExprMutator {
         if (dev_type->value == kDLCPU && storage_scope == "global") {
           int32_t constant_size = op->constant_allocation_size();
-          if (constant_size > 0 && constant_size * nbytes < runtime::kMaxStackAlloca) {
+          if (constant_size > 0 && constant_size * nbytes < kMaxStackAlloca) {
```
Maybe you could also change the `<` to `<=`.  Intuitively, an allocation of size `kMaxStackAlloca` should still be allowed on the stack, but the `<` would prevent it.

SebastianBoblestETAS replied:
I changed the code accordingly

manupa-arm commented on src/driver/driver_api.cc:
```diff
@@ -47,6 +47,7 @@ TVM_REGISTER_PASS_CONFIG_OPTION("tir.disable_vectorize", Bool);
 TVM_REGISTER_PASS_CONFIG_OPTION("tir.is_entry_func", Bool);
 TVM_REGISTER_PASS_CONFIG_OPTION("tir.add_lower_pass", Array<Array<ObjectRef>>);
 TVM_REGISTER_PASS_CONFIG_OPTION("tir.debug_keep_trivial_loop", Bool);
+TVM_REGISTER_PASS_CONFIG_OPTION("tir.max_stack_alloca", Integer);
```
Should we move this under namespace of the pass lower_tvm_builtin (e.g.  tir.lower_tvm_builtin.max_stack_alloca) ? and move it from driver_api into the pass implementation ?

SebastianBoblestETAS replied:
I looked into https://github.com/apache/tvm/blob/main/tests/python/unittest/test_tir_transform_lower_tvm_builtin.py
I am not quite sure how I would test that tir.max_stack_alloca works as expected. I would need to create an IRModule with AllocateNodes and check whether they are or are not in there after tranformations, depending on its value, right?
I do not really know how to do that unfortunately

manupa-arm replied:
Hi @SebastianBoblestETAS ,

I guess you are anwsering the the question, I have asked in the title.
Let me cite an similar example : 

https://github.com/apache/tvm/blob/6720d3593d4dac6015418d4b7e9ad875bbf0b0a2/tests/python/unittest/test_tir_transform_inject_double_buffer.py#L56-L60

We could use a similiar tvm.tir.stmt_functor.post_order_visit that counts the remaining Allocate nodes.

For this thread,

What do you think about moving the PassConfig option to be part of configuration of lower_tvm_builtin pass rather than here in the driver_api ?  (because this configuration now essential becomes a configuration of the pass)

SebastianBoblestETAS replied:
Hi, sorry that this took me so long. I moved the option to lover_tvm_builtin.cc as proposed. I also created a test for this but it fails with 
Check failed: (device_type_.defined()) is false: Unknown device type in current IR

How can I specify a device?

I did not yet push this to this branch here, but the changes I made are here:
https://github.com/SebastianBoblestETAS/tvm/compare/makeMaxStackAllocaConfigurable...SebastianBoblestETAS:makeMaxStackAllocaConfigurableFix

manupa-arm replied:
Hi @SebastianBoblestETAS ,

that information is usually inserted by Make(Un)PackedAPI transform and required by lower_builtin pass. Please take a look here : https://github.com/apache/tvm/blob/b365586159cd76a7fcd12d7915d2b59cc59e7105/tests/python/unittest/test_tir_transform_lower_tvm_builtin.py#L57-L68

manupa-arm reviewed (changes requested):
Thanks for the doing this @SebastianBoblestETAS  -- I did note this in the issue #9022 
Would it be possible to add a test here : https://github.com/apache/tvm/blob/main/tests/python/unittest/test_tir_transform_lower_tvm_builtin.py ?

junrushao1994 commented:
Any updates?

SebastianBoblestETAS commented:
Hi, I failed with writing a working unit test against this. I tried a few things and am currently stuck.

manupa-arm commented:
@SebastianBoblestETAS if it is the device issue,
running MakePackedAPI() pass (as shown in the example) before LowerTVMBuiltin should resolve the issue

SebastianBoblestETAS commented:
@manupa-arm Thanks for the hint. Sorry again that this is delayed so long. I will go back and look at it next week.

SebastianBoblestETAS commented:
Hi, 
I would postpone resolving this PR until we have finished our work the TIR export: https://discuss.tvm.apache.org/t/export-tir-to-json/12329/6 if this is ok.
Then, we could also simply write unit tests against the exported TIR to show that the max stack alloca option works correctly.

SebastianBoblestETAS commented:
Sorry that this stayed here so long. I just could not get a suitable test for this to work.
I close this now because our team actually has no more use for it.
If someone is interested in having this feature merged, I will of course work on it.
